background/adapter/qbittorrent.js

```diff
@@ -0,0 +1,168 @@
+'use strict';
+
+torrentToWeb.adapter.qbittorrent = function (baseUrl, username, password, autostart) {
+    let baseUrlObject = new URL(baseUrl);
+    baseUrlObject.pathname = '/api/v2';
+
+    baseUrl = baseUrlObject.toString();
+    let refUrl = baseUrl.replace(/(:\d+).*$/, '$1');
+    let filterUrls = refUrl.replace(/:\d+.*$/, '/*');
+    let sessionCookie = null;
+
+    function login () {
+        let requestData = new URLSearchParams();
+        requestData.append('username', username);
+        requestData.append('password', password);
+
+        return new Promise((resolve, reject) => {
+            fetch(baseUrl + '/auth/login', {
+                method: 'POST',
+                credentials: 'omit',
+                body: requestData,
+            }).then((response) => {
+                if (response.ok) {
+                    return response.text();
+                }
+
+                throw new Error(response.status.toString() + ': ' + response.statusText);
+            }).then((txt) => {
+                if (txt === 'Ok.') {
+                    resolve();
+                    return;
+                }
+
+                throw new Error('Login error: ' + txt);
+            }).catch((error) => reject(error));
+        });
+    }
+
+    function logout () {
+        return new Promise((resolve, reject) => {
+            fetch(baseUrl + '/auth/logout', {
+                method: 'POST',
+            }).finally(() => {
+                sessionCookie = null;
+                removeFilter();
+                resolve();
+            }).catch((error) => reject(error));
+        });
+    }
+
+    return {
+        send: function (filenameOrUrl, data, callback) {
+            let requestData = new FormData();
+
+            if (! autostart) {
+                requestData.append('paused', 'true');
+            }
+
+            if (filenameOrUrl.startsWith('magnet:')) {
+                installFilter();
+                login().then(() => {
+                    requestData.append('urls', filenameOrUrl + '\n');
```

Comment:
It seems like this can be simpified a lot. The other difference between magnet link and torrent is the append line. Thus you can just differ the append call via the if statement and leave the rest the same.

Reply:
Right. I'm blind ;-) ... Well, that's what reviews are for.
Just a moment, new commit almost on it's way...